### PR TITLE
pybind/mgr/dashboard: initial block integration

### PR DESCRIPTION
**Pull request by dillaman**
Signed-off-by: Jason Dillaman <dillaman@redhat.com>

### CODE REVIEW

File: src/pybind/mgr/dashboard/base.html [hunk 1]
```diff
@@ -97,6 +97,25 @@
                 return truncated_float + units[unit];
             };
 
+            rivets.formatters.dimless_binary = function(n){
```
Comment (jcsp) [inline src/pybind/mgr/dashboard/base.html]: Could take the common bits of this and dimless into a function that just takes the 1024/1000 and the list of suffixes?

File: src/pybind/mgr/dashboard/rbd_ls.py [hunk 2]
```diff
@@ -2,7 +2,47 @@
...
+                ioctx = cluster.open_ioctx(pool)
+                ioctx.stat("rbd_directory")
+                self.rbd_pools.append(pool)
+            except (rados.PermissionError, rados.ObjectNotFound):
+                self.log.info("No RBD directory in " + pool)
+            except:
+                self.log.error("Failed to open pool " + pool)
+        return self.rbd_pools
+
+class RbdPools(VersionedSyncObject):
```
Comment (jcsp) [inline src/pybind/mgr/dashboard/rbd_ls.py]: The whole VersionedSyncObject is a bit of a red herring here, I should rip it out.  It's a hangover from Calamari when there was a layer beneath that used the version info to work out when something fresh needed fetching.

Comment (dillaman) [inline src/pybind/mgr/dashboard/rbd_ls.py]: Yeah -- I couldn't for the life of me figure out how it was being being used. I just figured it was a "TODO" so if I could just add code to cache RBD pools per epoch, it's one less thing to change if and when it does cache.

File: src/pybind/mgr/dashboard/rbd_ls.py [hunk 3]
```diff
@@ -2,7 +2,47 @@
...
+        super(RbdPoolLs, self).__init__(module_inst)
+        self.osd_pools = osd_pools
+        self.rbd_pools = []
+
+    def _get(self):
+        from mgr_module import ceph_state
+        ctx_capsule = ceph_state.get_context()
+
+        self.log.info("Constructing Rados")
+        cluster = rados.Rados(context=ctx_capsule)
```
Comment (jcsp) [inline src/pybind/mgr/dashboard/rbd_ls.py]: when I wrote the rbd_ls.py stuff it was the only place spinning up a rados.Rados -- now that we're using it elsewhere, the instance should probably be shared

**Comment by jcsp** [comment]
I think the idea of scanning for rbd-using pools per OSD epoch is probably overkill, and also not 100% watertight because if I create a pool it's initially not an rbd pool, it's only when I write some stuff that it becomes one (and of course writing some stuff doesn't affect the epoch).

We could have a class that only updates itself when it either sees new pools in the osdmap (it can read that every time since osdmap is right here in memory), or if there was a pool which previously had zero objects and now has >0 objects.  That would cover all the "normal" cases, although it could still be fooled by someone who e.g. repurposed a cephfs data pool for use with rbd without telling us.

I kind of left this out when doing the quickie rbd_ls, because I wasn't sure whether we would do something like this first, or whether we would do something that let us explicitly tag pools in the osdmap with what subsystem was going to use them.

**Comment by liewegas** [comment]
What about a 'tag' concept for pools, stored in the osdmap.  It could be a simple set<string> in pg_pool_t and would make it super easy to identify pools for things like this.

rbd could refuse to operate on a pool not tagged with 'rbd', 'ceph fs new' would add 'cephfs' tags, etc.

The downside is we are going from 0 steps before 'rbd create' to

- ceph osd pool create rbd N
- ceph osd pool tag rbd rbd

**Comment by dillaman** [comment]
@jcsp @liewegas RBD currently creates and deletes a self-managed snapshot if the "rbd_directory" object is missing, so that is why I was using the epoch for tracking. As for tags, the initial "rbd create" on the pool could easily set a tag if unset.

**Comment by jcsp** [comment]
Ah, I did not realise that there was a snapshot when rbd_directory was created.

Setting the tag during rbd create would need higher caps than an rbd admin normally needs to just write to OSDs, but I imagine in many (most?) cases it's really going to be someone using client.admin key to run the rbd CLI anyway.

**Comment by jcsp** [comment]
I've reworked a bit here if you want to cherry-pick it: https://github.com/ceph/ceph/commit/df56311b9fb3033af7b1899849afbb91d087baef

This depends on https://github.com/ceph/ceph/pull/15577

**Comment by dillaman** [comment]
@jcsp update pushed

File: src/pybind/mgr/dashboard/module.py [hunk 4]
```diff
@@ -376,6 +380,20 @@ def _toplevel_data(self):
                 """
                 Data consumed by the base.html template
                 """
+                osd_map = global_instance().get_sync_object(OsdMap).data
```
Comment (jcsp) [inline src/pybind/mgr/dashboard/module.py]: unused instance of osd_map

**Pull request merged**